lib/exercises.rb

```diff
@@ -2,19 +2,67 @@
 
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n) where n is the number of strings for the hashing function but I'm not sure how split/sort/join factors in (maybe O of m where m is the number of letters in all the strings?). 
+# Space Complexity: O(n) because we are creating a new place in memory for each string
 
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+
+  hash = {}
+
+  # take each word in 'strings' array ...
+  strings.each do |word|
+    # ... and split it into letters, alphabetize those letters, and join them back together: this produces a standardized string for words that are anagrams of each other ('ate', 'eat' => 'aet')
+    letters = word.split(//).sort().join()
+    
+    # if the standardized string isn't already in the hash, add it as the key and an empty array as the value
+    if hash[letters].nil?
+      hash[letters] = []
+    end
+
+    # push the word to the array for the associated standardized string
+    hash[letters] << word
+
+  end
+
+  # return the values (array of words that are anagrams) for each standardized string
+  return hash.values
+
 end
 
 # This method will return the k most common elements
 # in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n*2) => O(n): one O(n) pass to create the hash map and a second to determine max values
+# Space Complexity: O(n) 
 def top_k_frequent_elements(list, k)
```

Comment:
You do have to read the documentation for `max_by` it's actually going to sort the values to find the max k elements, so... this is actually O(n log n) where n is the maximum number of elements.